test/spec/automatic/tagspec.js

```diff
@@ -77,10 +77,232 @@ describe("Cloudinary.ImageTag", function() {
+    let breakpointListFloat = [...breakpointList];
+    breakpointListFloat.forEach(function (part, index, theArray) {
+      theArray[index] += 0.1;
+    });
+    let commonSrcsetFloat = {'breakpoints': breakpointListFloat};
+
+    let tag = new cloudinary.ImageTag(
+        fullPublicId,
+        Object.assign({}, commonImageOptions, {srcset: commonSrcsetFloat})
+    ).toHtml();
+    let expectedTag = expectedImageTagFromParams(fullPublicId, commonTransformationStr, '', breakpointListFloat);
+
+    return expect(expectedTag).toBe(tag);
+  });
+
+  it("Should support srcset attribute defined by min_width, max_width, and max_images", function () {
+    let srcsetParams = {'min_width': minWidth, 'max_width': maxWidth, 'max_images': breakpointList.length};
+
+    let tag = new cloudinary.ImageTag(
+        fullPublicId,
+        Object.assign({}, commonImageOptions, {srcset: srcsetParams})
+    ).toHtml();
+    let expectedTag = expectedImageTagFromParams(fullPublicId, commonTransformationStr, '', breakpointList);
+
+    return expect(expectedTag).toBe(tag);
+  });
+
+  it("Should support 1 image in srcset", function () {
+    let srcsetParams = {'min_width': minWidth, 'max_width': maxWidth, 'max_images': 1};
+    let srcsetBreakpoint = {"breakpoints": [maxWidth]};
+
+    let expectedTag = expectedImageTagFromParams(fullPublicId, commonTransformationStr, '', [maxWidth]);
+    let tagByParams = new cloudinary.ImageTag(
+        fullPublicId,
+        Object.assign({}, commonImageOptions, {srcset: srcsetParams})
+    ).toHtml();
+    let tagByBreakpoint = new cloudinary.ImageTag(
+        fullPublicId,
+        Object.assign({}, commonImageOptions, {srcset: srcsetBreakpoint})
+    ).toHtml();
+
+    expect(expectedTag).toBe(tagByParams);
+    return expect(expectedTag).toBe(tagByBreakpoint);
+  });
+
+  it("Should support custom transformation for srcset items", function () {
+    let srcsetParams = Object.assign({}, commonSrcset, {
+      transformation: {'crop': 'crop', 'width': 10, 'height': 20},
+    });
+    let customTransformationStr = 'c_crop,h_20,w_10';
+
+    let expectedTag = expectedImageTagFromParams(
+        fullPublicId,
+        commonTransformationStr,
+        customTransformationStr,
+        breakpointList
+    );
+    let tag = new cloudinary.ImageTag(
+        fullPublicId,
+        Object.assign({}, commonImageOptions, {srcset: srcsetParams})
+    ).toHtml();
+
+    return expect(expectedTag).toBe(tag);
+  });
+
+  it("Should populate sizes attribute", function () {
+    let srcsetParams = Object.assign({}, commonSrcset, {
+      sizes: true,
+    });
+    let expectedSizes = [];
+    for(let key in breakpointList){
+      if (breakpointList.hasOwnProperty(key)) {
+        let w = breakpointList[key];
+        expectedSizes.push(`(max-width: ${w}px) ${w}px`);
+      }
+    }
+    let attributes = {"sizes": expectedSizes.join(', ')};
+
+    let expectedTag = expectedImageTagFromParams(
```

Comment:
To be honest I really appreciate the reuse of the `expectedImageTagFromParams`, however I think this makes the tests very difficult to read.

It's an abstraction that makes it difficult to understand what's really happening, without running mentally the code within `expectedImageTagFromParams`.

this is similar to something like this:
```js
expect(555+666).toBe(generateCalculation(555, 666)) 
```

The point is that 
```js
expect(1+1).toBe(2)
```

Will forever be more readable.
@maoznir what do you think?

Reply:
I agree, it's also somewhat reimplementing part of the sdk in the test.
I see something similar here but much more readable: https://github.com/cloudinary/cloudinary_gem/blob/master/spec/image_spec.rb
But even then I would prefer that each test will include the wanted result in plain text